02-microarray/pathway-analysis_microarray_03_gsea.Rmd

```diff
@@ -0,0 +1,343 @@
+    keys = dge_df$Gene,
+    column = "ENTREZID", # Replace with the type of gene identifiers you would like to map to
+    keytype = "ENSEMBL", # Replace with the type of gene identifiers in your data
+    multiVals = "first" # This will keep only the first mapped value for each Ensembl ID
+  )
+) %>%
+  # If an Ensembl gene identifier doesn't map to a Entrez gene identifier, drop that
+  # from the data frame
+  dplyr::filter(!is.na(entrez_id)) %>%
+  # Make an `Ensembl` column to store the rownames
+  tibble::rownames_to_column("Ensembl") %>%
+  # Now let's join the rest of the expression data
+  dplyr::inner_join(dge_df, by = c("Ensembl" = "Gene"))
+```
+
+This `1:many mapping between keys and columns` message means that some Ensembl gene identifiers map to multiple Entrez IDs.
+In this case, it's also possible that a Entrez ID will map to multiple Ensembl IDs.
+For the purpose of performing GSEA later in this notebook, we keep only the first mapped IDs.
+For more about how to explore this, take a look at our [microarray gene ID conversion example](https://alexslemonade.github.io/refinebio-examples/02-microarray/gene-id-annotation_microarray_01_ensembl.html).
+
+Let's see a preview of `dge_mapped_df`.
+
+```{r}
+head(dge_mapped_df)
+```
+
+Now let's check to see if we have any Entrez IDs that mapped to multiple Ensembl IDs.
+
+```{r}
+any(duplicated(dge_mapped_df$entrez_id))
+```
+
+Looks like we do have duplicated Entrez IDs.
+Let's find out which Entrez IDs have been duplicated.
+
+```{r}
+dge_mapped_df %>%
+  dplyr::filter(duplicated(entrez_id)) %>%
+  dplyr::pull(entrez_id)
+```
+
+Now let's take a look at the rows associated with one of the duplicated Entrez IDs.
+
+```{r}
+dge_mapped_df %>%
+  dplyr::filter(entrez_id == "336702")
+```
+We can see that the associated values vary for each row although the Entrez ID is the same.
+
+We do not want duplicated gene identifiers for the GSEA steps later, so let's keep the Entrez IDs associated with the higher t-statistic value.
+
+```{r}
+filtered_dge_mapped_df <- dge_mapped_df %>%
+  # Sort so that highest t-statistic values are at the top
+  dplyr::arrange(dplyr::desc(t)) %>%
+  # Filter out the duplicated rows using `dplyr::distinct()`-- this will keep
+  # the first row with the duplicated value thus keeping the row with the
+  # highest t-statistic value
+  dplyr::distinct(entrez_id, .keep_all = TRUE)
+```
+
+Let's check to see that we removed the duplicate Entrez IDs and kept the row with the higher t-statistic value.
+
+```{r}
+filtered_dge_mapped_df %>%
+  dplyr::filter(entrez_id == "336702")
+```
+
+Looks like we were able to successfully get rid of the duplicate gene identifiers and keep the observation with the higher t-statistic value!
+Note however, that a caveat in using this approach is that the genes that have duplicate identifiers could be enriched in a particular pathway/gene set and we may get an overly optimistic view of how perturbed that pathway truly is.
```

Comment:
Move to ~line 296 and _expand upon_. At this point, you haven't explained GSEA yet (https://github.com/AlexsLemonade/refinebio-examples/pull/339/files#diff-0d0fd22250c391b41f655f78697b0bcbd9626a54072e31fc31cbd01c6faf295dR295)! So I'd also think about, now that we need to explain scientific decision making during the gene identifier conversion, if we want to move up that explanation to before the gene ID conversion.

Reply:
I moved this line up to ~line 296 in this PR and left a comment on #347 to address the latter part of this comment: https://github.com/AlexsLemonade/refinebio-examples/pull/347#discussion_r518297547